### PR TITLE
Selection improvements for the capture window.

### PR DESCRIPTION
**Pull request by vickyliu-go4it**
Enable highlighting selected function calls in both the capture window and the live functions data view panel.
- When picking an interested function in the live function data view panel, all calls of the same function will be highlighted in the capture window.
- When picking an interested function call in the capture window, the corresponding function in the live function data view panel will be highlighted.

Fixes b/168104538.

### CODE REVIEW

File: OrbitGl/TimerTrack.cpp
Comment (pierricgimmig) [inline OrbitGl/TimerTrack.cpp:111]: Is this needed? kAllThreadsFakeTid is used only for sampling events if I'm not mistaken so we shouldn't have to check for it when drawing timers right?

File: OrbitGl/TimerTrack.cpp
Comment (pierricgimmig) [inline OrbitGl/TimerTrack.cpp:112]: This looks like it's only highlighting timers on the selected thread, shouldn't we highlight on every thread?

File: OrbitQt/orbitglwidget.cpp
Comment (pierricgimmig) [inline OrbitQt/orbitglwidget.cpp:41]: the check here is performed on every event, it should probably be done only on the paint event. Actually, I find this is a bit convoluted, could we simplify by keeping track of the canvas' selected function address, and running the highlight logic when we detect it has changed?

File: OrbitQt/orbitmainwindow.cpp
Comment (pierricgimmig) [inline OrbitQt/orbitmainwindow.cpp:750]: Why is the check against kAllThreadFakeTid needed?

File: OrbitGl/TimerTrack.cpp
Comment (vickyliu-go4it) [reply OrbitGl/TimerTrack.cpp:111]: Thanks for pointing out this. I misunderstood the purpose for checking kAllThreadsFakeTid. I thought this is used to filter out some uninterested thread. Thanks for explaining this to me : )

File: OrbitGl/TimerTrack.cpp
Comment (vickyliu-go4it) [reply OrbitGl/TimerTrack.cpp:112]: Thanks for the suggestion!

File: OrbitGl/TimerTrack.cpp
Comment (vickyliu-go4it) [reply OrbitGl/TimerTrack.cpp:115]: The first statement is for the case that we pick a textbox in the capture window, and the second one is for the case that we pick a function from the live function data view panel. But this might not be a good coding style. Thanks for pointing out this. I will change to use parenthesis for distinguishing between those two cases.

File: OrbitGl/CaptureWindow.cpp
Comment (vickyliu-go4it) [reply OrbitGl/CaptureWindow.cpp:171]: Great suggestion! Thanks so much.

File: OrbitQt/orbitglwidget.cpp
Comment (vickyliu-go4it) [reply OrbitQt/orbitglwidget.cpp:41]: Totally agree and thanks for pointing out this. I had a meant to do this check for every mouse events that might incur the selection of a textbox, but I didn't find a suitable if condition. Thanks for explaining this to me.

**Review by karupayun — COMMENTED** [review]
Hey vicky, thanks for addressing this issue :)

In general looks very good, but please notice (it's also in the comments):
1- If you use the arrow_keys, the selection in LiveFunctionData is not updated.
2- If you select some function in Queue tracks, all functions there are selected.

File: OrbitGl/GlCanvas.h
Comment (vickyliu-go4it) [reply OrbitGl/GlCanvas.h:188]: Thanks for that.

File: OrbitGl/LiveFunctionsDataView.cpp
Comment (karupayun) [reply OrbitGl/LiveFunctionsDataView.cpp:414]: Or another option is using find_if function, as Anton is making here:
https://github.com/google/orbit/pull/1505/files#diff-f42037594387cf3f40c3efa4aeb67ae35e16ba6dbcb856eeb9263f6339f76be5R145

File: OrbitGl/App.cpp
Comment (vickyliu-go4it) [reply OrbitGl/App.cpp:1391]: Thanks for pointing out the "all timers highlighted issue". I set highlighted_function as 0 to indicate no highlighted function is picked from the live functions data view panel. You remind me that I should also check the function address of selected text box because when text box is selected from places such as Transfer Queue, the function address is 0.

File: OrbitGl/App.cpp
Comment (vickyliu-go4it) [reply OrbitGl/App.cpp:1377]: Good point! Thanks.

File: OrbitGl/DataManager.cpp
Comment (vickyliu-go4it) [reply OrbitGl/DataManager.cpp:53]: Thanks.

File: OrbitGl/LiveFunctionsDataView.cpp
Comment (vickyliu-go4it) [reply OrbitGl/LiveFunctionsDataView.cpp:414]: Yes, exactly! Thanks for that.

File: OrbitGl/LiveFunctionsDataView.h
Comment (vickyliu-go4it) [reply OrbitGl/LiveFunctionsDataView.h:29]: Noted with thanks!

File: OrbitQt/orbitmainwindow.cpp
Comment (vickyliu-go4it) [reply OrbitQt/orbitmainwindow.cpp:755]: Sure, exactly. Thanks for pointing out this.

File: OrbitQt/orbitmainwindow.cpp
Comment (vickyliu-go4it) [reply OrbitQt/orbitmainwindow.cpp:758]: Totally agree. Thanks!

File: OrbitQt/orbitlivefunctions.cpp
Comment (vickyliu-go4it) [reply OrbitQt/orbitlivefunctions.cpp:112]: Yes, make sense. A CHECK would be much better.

File: OrbitQt/orbitmainwindow.cpp
Comment (vickyliu-go4it) [reply OrbitQt/orbitmainwindow.cpp:750]: Thanks for pointing out this. I misunderstood the purpose for checking against kAllThreadFakeTid. I thought this will filter out some uninterested thread id. I will read the OrbitGl codes more carefully.

File: OrbitQt/orbitmainwindow.cpp
Comment (vickyliu-go4it) [reply OrbitQt/orbitmainwindow.cpp:212]: > Please use new syntax (https://wiki.qt.io/New_Signal_Slot_Syntax)

Thanks for that. Change to use the Qt5 syntax.

Comment (vickyliu-go4it) [reply OrbitQt/orbitmainwindow.cpp:212]: > Actually are the signal and slot really needed here?

Sorry that I don't have a good answer for this at present. I have no idea about other methods to trigger checking and changing of function highlight. So here I use signal and slot. Every time when we select a textbox by picking with mouse or the arrow keys from the capture window, I want to inform the live functions data view panel to change the selection. I will search for a better solution. Thanks for raising this issue.

File: OrbitGl/App.cpp
Comment (karupayun) [reply OrbitGl/App.cpp:1391]: There are some Thread Fake ID, like -1 or -2, I don't remember. If you declare a constant kNotUsedThreadID or whatever (with a value of -3 for example), you can set the value to this ID and you can avoid the double check

Comment (vickyliu-go4it) [reply OrbitGl/App.cpp:1391]: Thanks so much for telling me about the details of Thread Fake ID. But I'm not quite understand what the 'double check' refers to. Do you mean the check of uninterested threads?
BTW, I fixed the "all timers highlighted issue" in the fixup commit a654230 by  checking the function address of selected textbox  when declaring `need_highlight`.

Comment (karupayun) [reply OrbitGl/App.cpp:1391]: Ignore my double check comment. I misunderstood something. Anyway, if you still want to set 0 there, you should create a constant for that.

File: OrbitGl/TimerTrack.cpp
Comment (karupayun) [inline OrbitGl/TimerTrack.cpp:111]: Maybe a better way to implement this is by making variables.
bool picked_from_capture_windows = ....
bool picked_from_live_func_panel = ....
bool need_highlight = picked_from_capture_windows || picked_from_live_func_panel 

And you don't need comments. What do you think?

File: OrbitGl/TimerTrack.cpp
Comment (karupayun) [inline OrbitGl/TimerTrack.cpp]: If you are still leaving the 0, I think that we should have a comment there explaining our assumption that ThreadID are greater than 0. Still, for me is better, if possible, to have "GOrbitApp->highlighted_function() != kFakeTid"

**Review by karupayun — APPROVED** [review]
LGTM! :) :)

I've left 2 comments. And please wait for Pierric approval before merged.

Maybe for future PR's, did you think what we should highlight when there are more than 1 selected function in LiveFunction tab?

File: OrbitGl/TimerTrack.cpp
Comment (vickyliu-go4it) [reply OrbitGl/TimerTrack.cpp]: Thanks for the suggestion. I chaged to use a constant named kUnusedHighlightedFunctionAddress in the latest fixup commit db98890. But because of my bad naming of those fixup commits, it's hard to notice the purpose of each fixup.

File: OrbitGl/TimerTrack.cpp
Comment (vickyliu-go4it) [reply OrbitGl/TimerTrack.cpp:111]: Good suggest! Will do that : )

**Comment by pierricgimmig** [comment]
Hi Tingwei, thanks for the changes, trying out the change and I noticed a bug:

image

becomes:

image

The manually instrumented scopes are all highlighted. You can reproduce by profiling OrbitTest for example, which is compiled by default and is present next to the OrbitService executable.

**Comment by vickyliu-go4it** [comment]
> Hi Tingwei, thanks for the changes, trying out the change and I noticed a bug:
> 
> image
> 
> becomes:
> 
> image
> 
> The manually instrumented scopes are all highlighted. You can reproduce by profiling OrbitTest for example, which is compiled by default and is present next to the OrbitService executable.

Hi Pierric,

Thanks for helping me test the change. I have profiled OrbitTest. This is not a bug. I have output the function addresses of the highlighted function and the clicked text box. They are both correct.

When picking the 1st function in the live Tab or picking the corresponding text boxes in the capture window,  not all manually instrumented functions are highlighted. Only the functions whose addresses are the same with the function picked in the capture window or the live Tab are highlighted.

Screenshot 2020-12-03 225510

When picking the 2nd function in the live Tab or picking the corresponding text boxes in the capture window, still only functions with matching function addresses will be highlighted.



It is true that under certain condition, many manually instrumented functions are highlighted. But this case is normal, because all the highlighted functions have the same address. (see the following 3 figures)




Here is one more example to show many functions are highlighted, but all their function addresses match our selected function addresses.



Thanks again for helping me test the code change.

**Comment by pierricgimmig** [comment]
Hi Tingwei, 

Thanks for looking into it. I think highlighting all manually instrumented functions could be misleading.

I looked into it a bit more on my side, please see suggestions in https://github.com/pierricgimmig/orbit/commit/6cc69da06e5142d4c1e275a4a63067d815b66a77. 

Feel free to incorporate in your PR.

**Pull request merged**
